lib/improver/utilities/cube_extraction.py

```diff
@@ -0,0 +1,111 @@
+# (C) British Crown Copyright 2017 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Utilities to parse a list of constraints and extract matching subcube """
+
+from ast import literal_eval
+import iris
+
+iris.FUTURE.netcdf_no_unlimited = True
+iris.FUTURE.netcdf_promote = True
+
+
+def parse_constraint_list(constraints, units):
+    """
+    Takes a list of string constraints and converts to key-value pairs
+
+    Args:
+        constraints (list):
+            Space separated list of constraints with no space between key and
+            value in each pair: e.g: kw1=val1 kw2=val2 kw3=val3.  Values must
+            be of scalar types interpretable by ast.literal_eval: strings,
+            numbers, booleans, or "None".
+        units (list of strings or None):
+            Space separated list of units for each coordinate in the list of
+            constraints.  One or more "units" may be None, and units can only
+            be associated with coordinate constraints.
+
+    Returns a dictionary of constraints and units
+    """
+
+    if units is None:
+        list_units = len(constraints)*[None]
+        units_dict = None
+    else:
+        if len(units) != len(constraints):
+            msg = 'units list must match constraints'
+            raise ValueError(msg)
+        list_units = units
+        units_dict = {}
+
+    constraints_dict = {}
+    for constraint_pair, units in zip(constraints, list_units):
+        [key, value] = constraint_pair.split('=')
```

Comment:
The formatting requirements for input to this CLI are quite strict. ```time=10``` is fine but ```time = 10``` is not. I wonder if there is a slightly better way of capturing the input to make it a bit more robust? See perhaps improver-nbhood --radii as a potential alternative. I don't think we have to change this, but it might be worth a bit of thought just to see if a nicer solution presents itself to you.

Reply:
With `args.constraints` having `nargs='+'` in the main CLI, we can pass in a collection of space separated constraints:
`improver extract infile.nc outfile.nc "const1 = 1" "const2=2"` => `args.constraints = ['const1 = 1', 'const2=2']`
..so I guess we could split on `=`, and then apply `strip()` to both items in the resulting list.

Is that better? (I haven't looked at improver-nbhood though). Or we could just have constraints as a string, and separate by something other than a space (commas?).